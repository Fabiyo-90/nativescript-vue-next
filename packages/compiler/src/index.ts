import {
    baseCompile,
    baseParse,
    CompilerOptions,
    CodegenResult,
    ParserOptions,
    RootNode,
} from '@vue/compiler-core'

export function compile(
    template: string,
    options: CompilerOptions = {}
): CodegenResult {
    return baseCompile(template, {
        ...options,
        nodeTransforms: [],
        directiveTransforms: {
            ...(options.directiveTransforms || {})
        },
        // transformHoist: stringifyStatic
    })
}

export function parse(template: string, options: ParserOptions = {}): RootNode {
    return baseParse(template, {
        // ...parserOptions,
        ...options
    })
}

// export * from './runtimeHelpers'
// export { transformStyle } from './transforms/transformStyle'
// export { createDOMCompilerError, DOMErrorCodes } from './errors'
export * from '@vue/compiler-core'
