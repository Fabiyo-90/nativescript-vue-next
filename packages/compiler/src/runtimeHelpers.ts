import { registerRuntimeHelpers } from '@vue/compiler-core'

export const V_MODEL = Symbol(__DEV__ ? `vModel` : ``)

export const V_ON_WITH_MODIFIERS = Symbol(__DEV__ ? `vOnModifiersGuard` : ``)
export const V_ON_WITH_KEYS = Symbol(__DEV__ ? `vOnKeysGuard` : ``)

export const V_SHOW = Symbol(__DEV__ ? `vShow` : ``)

export const TRANSITION = Symbol(__DEV__ ? `Transition` : ``)
export const TRANSITION_GROUP = Symbol(__DEV__ ? `TransitionGroup` : ``)

// todo: we might want to move built-in components into 'resolveAssets'
// as we are already treating all elements as "components" and resolving them
// to components/actual elements at runtime
export const ACTION_BAR = Symbol(__DEV__ ? `ActionBar` : ``)

registerRuntimeHelpers({
  [V_MODEL]: `vModel`,
  [V_ON_WITH_MODIFIERS]: `withModifiers`,
  [V_ON_WITH_KEYS]: `withKeys`,
  [V_SHOW]: `vShow`,
  [TRANSITION]: `Transition`,
  [TRANSITION_GROUP]: `TransitionGroup`,
  [ACTION_BAR]: `ActionBar`,
})
