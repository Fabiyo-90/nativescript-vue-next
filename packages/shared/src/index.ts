import { Trace } from '@nativescript/core'

type DebugCategory = string | typeof Trace.categories

export function debug(
  s: any,
  category: DebugCategory = Trace.categories.Debug
) {
  Trace.write(s, category as string, Trace.messageType.log)
}

// const onRE = /^on[^a-z]/
const onRE = /^on.+/
export const isOn = (key: string) => onRE.test(key)
