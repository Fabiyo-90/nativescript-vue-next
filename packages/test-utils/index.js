'use strict'

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./dist/test-utils.cjs.prod.js')
} else {
  module.exports = require('./dist/test-utils.cjs.js')
}
