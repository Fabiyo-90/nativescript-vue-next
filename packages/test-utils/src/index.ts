export * from './mocks/core/ViewBase.mock'
export * from './mocks/core/Frame.mock'
export * from './mocks/registry.mock'
export * from './platform'
export * from './mockWarn'
