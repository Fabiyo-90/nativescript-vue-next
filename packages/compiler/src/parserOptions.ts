import {
  TextModes,
  ParserOptions,
  ElementNode,
  Namespaces,
  isBuiltInType
  // NodeTypes,
  // isBuiltInType
} from '@vue/compiler-core'
import { makeMap } from '@vue/shared'
// import { isKnownView } from '@nativescript-vue/runtime'
import namedCharacterReferences from './namedChars.json'
import { ACTION_BAR, TRANSITION, TRANSITION_GROUP } from './runtimeHelpers'

const isRawTextContainer = /*#__PURE__*/ makeMap('style,script', true)

export const enum DOMNamespaces {
  HTML = Namespaces.HTML
}

const isBuiltInComponent = (tag: string): symbol | undefined => {
  if (isBuiltInType(tag, `ActionBar`)) {
    return ACTION_BAR
  } else if (isBuiltInType(tag, `Transition`)) {
    return TRANSITION
  } else if (isBuiltInType(tag, `TransitionGroup`)) {
    return TRANSITION_GROUP
  }
}

export const parserOptions: ParserOptions = {
  // We don't have void tags in NativeScript (<br>, <hr> etc.)
  isVoidTag: tag => false,
  isNativeTag: tag => {
    if (isBuiltInComponent(tag)) {
      return false
    }
    // in case we encounter an element that is not a built-in component
    // we will assume it's a plugin provided element
    return true
  },
  isPreTag: tag => tag === 'pre',
  isBuiltInComponent,
  // todo: we might add support for different namespaces in the future
  // for example - it would be neat to be able to write inline svg
  // with the svg plugin?
  getNamespace(tag: string, parent: ElementNode | undefined): DOMNamespaces {
    return DOMNamespaces.HTML
  },

  // https://html.spec.whatwg.org/multipage/parsing.html#parsing-html-fragments
  getTextMode(tag: string, ns: DOMNamespaces): TextModes {
    // todo: look into if any of this would be useful in ns
    if (ns === DOMNamespaces.HTML) {
      if (tag === 'textarea' || tag === 'title') {
        return TextModes.RCDATA
      }
      if (isRawTextContainer(tag)) {
        return TextModes.RAWTEXT
      }
    }
    return TextModes.DATA
  },

  // https://html.spec.whatwg.org/multipage/named-characters.html#named-character-references
  namedCharacterReferences,
  maxCRNameLength: /*#__PURE__*/ Object.keys(namedCharacterReferences).reduce(
    (max, name) => Math.max(max, name.length),
    0
  )
}
