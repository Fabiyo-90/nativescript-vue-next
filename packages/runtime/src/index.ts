import {
  Renderer,
  createRenderer,
  RootRenderFunction,
  CreateAppFunction,
  ComponentPublicInstance,
  VNode,
} from '@vue/runtime-core'
import { Application } from '@nativescript/core'
import { debug } from '@nativescript-vue/shared'
import { nodeOps } from './nodeOps'
import { patchProp } from './patchProp'
import { INSVElement, NSVRoot } from './nodes'
import { install as NavigationPlugin } from './plugins/navigation'
import './registry'

const rendererOptions = {
  patchProp,
  ...nodeOps,
}

let renderer: Renderer

function ensureRenderer() {
  return renderer || (renderer = createRenderer(rendererOptions))
}

function runApp(root: ComponentPublicInstance): ComponentPublicInstance {
  Application.run({
    create: () => {
      debug(
        `Root Node: ${JSON.stringify({
          id: root.$el.nodeId,
          type: root.$el.nodeType,
          tag: root.$el.tagName,
        })}`
      )
      return root.$el.nativeView
    },
  })

  return root
}

export const render = ((
  vnode: VNode | null,
  container: INSVElement | NSVRoot
) => {
  ensureRenderer().render(vnode, container)
}) as RootRenderFunction<INSVElement | NSVRoot>

export const createApp = ((...args) => {
  const app = ensureRenderer().createApp(...args)
  const { mount } = app

  app.mount = (): any => {
    return runApp(mount(nodeOps.createRoot()))
  }

  // Built-in plugins
  app.use(NavigationPlugin)

  return app
}) as CreateAppFunction<INSVElement>

export * from './nodeOps'
export * from './nodes'
export * from './runtimeHelpers'
export * from './registry'
export { resolveComponent } from './resolveAssets'

// Plugins
export { $navigateTo } from './plugins/navigation'

// runtime directive helpers
export { vModel } from './directives/vModel'
export { vShow } from './directives/vShow'
export { ListView } from './components/ListView'

// Runtime components
export { ActionBar } from './components/ActionBar'

export * from '@vue/runtime-core'

// TODO: remove the following export sentence when
// https://github.com/vuejs/vue-next/issues/1109 was fixed
export {
  // @ts-ignore
  registerRuntimeCompiler,
} from '@vue/runtime-core'
