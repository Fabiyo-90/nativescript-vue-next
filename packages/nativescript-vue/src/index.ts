//
import {
  compile,
  CompilerOptions,
  CompilerError
} from '@nativescript-vue/compiler'
import {
  registerRuntimeCompiler,
  RenderFunction,
  warn
} from '@nativescript-vue/runtime'
import * as runtime from '@nativescript-vue/runtime'
import { isString, NOOP } from '@vue/shared'

const compileCache: Record<string, RenderFunction> = Object.create(null)

function compileToFunction(
  template: string,
  options?: CompilerOptions
): RenderFunction {
  if (!isString(template)) {
    __DEV__ && warn('invalid template option: ', template)
    return NOOP
  }

  const key = template
  const cached = compileCache[key]

  if (cached) {
    return cached
  }

  const { code } = compile(template, {
    hoistStatic: true,
    onError(err: CompilerError) {
      // todo
    },
    ...options
  })

  const render = new Function('Vue', code)(runtime) as RenderFunction
  return (compileCache[key] = render)
}

registerRuntimeCompiler(compileToFunction)

export { compileToFunction as compile }
export * from '@nativescript-vue/runtime'
