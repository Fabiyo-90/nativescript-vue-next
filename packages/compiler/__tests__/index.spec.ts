import { compile } from '../src'
import { registerTestElement } from '@nativescript-vue/test-utils'

registerTestElement('Frame')
registerTestElement('Page')
registerTestElement('ActionBar')
registerTestElement('ActionItem')
registerTestElement('StackLayout')
registerTestElement('Label')

describe('compile', () => {
  // This is just a temporary test to see that the compiler actually compiles the template
  // to the correct blocks/render calls.
  it('works', () => {
    const { code } = compile(`<Frame>
    <Page>
        <ActionBar>
            <ActionItem v-if="foo" text="Test Item" />
            <ActionItem v-else text="Test Item" />
        </ActionBar>
        <StackLayout>
            <!-- Comment -->
            <Label v-for="i in 10" :key="i" @tap="onTap" :text="msg" ref="test" />
            <UnknownElement/>
        </StackLayout>
    </Page>
</Frame>`)

    expect(code).toMatchSnapshot()
  })
})
