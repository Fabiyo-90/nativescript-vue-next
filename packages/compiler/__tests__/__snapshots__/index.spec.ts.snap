// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile works 1`] = `
"const _Vue = Vue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { resolveComponent: _resolveComponent, createVNode: _createVNode, openBlock: _openBlock, createBlock: _createBlock, createCommentVNode: _createCommentVNode, renderList: _renderList, Fragment: _Fragment } = _Vue

    const _component_ActionItem = _resolveComponent(\\"ActionItem\\")

    return (_openBlock(), _createBlock(\\"Frame\\", null, [
      _createVNode(\\"Page\\", null, [
        _createVNode(\\"ActionBar\\", null, [
          foo
            ? (_openBlock(), _createBlock(_component_ActionItem, {
                key: 0,
                text: \\"Test Item\\"
              }))
            : (_openBlock(), _createBlock(_component_ActionItem, {
                key: 1,
                text: \\"Test Item\\"
              }))
        ]),
        _createVNode(\\"StackLayout\\", null, [
          (_openBlock(true), _createBlock(_Fragment, null, _renderList(10, (i) => {
            return (_openBlock(), _createBlock(\\"Label\\", {
              key: i,
              onTap: onTap,
              text: msg,
              ref: \\"test\\"
            }, null, 40 /* PROPS, HYDRATE_EVENTS */, [\\"onTap\\", \\"text\\"]))
          }), 128 /* KEYED_FRAGMENT */))
        ])
      ])
    ]))
  }
}"
`;
